[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1FI5iO_RAZXsyzA1OjinDzu59Y9q5ZZ-b?usp=drive_link)

## Introduction to Sentence Transformers

First, we install the `sentence-transformers` library:

In [1]:
!pip install sentence-transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 227.1/227.1 kB 2.0 MB/s eta 0:00:00
  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
  Using cached nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl (176.2 MB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-m

Next, we create our Sentence Transformer model. We will use the original SBERT model `bert-base-nli-mean-tokens`:

In [2]:
from sentence_transformers import SentenceTransformer
model = SentenceTransformer('bert-base-nli-mean-tokens')
model

/usr/local/lib/python3.10/dist-packages/sentence_transformers/cross_encoder/CrossEncoder.py:11: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/3.99k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/625 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/399 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/2.00 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

SentenceTransformer(
  (0): Transformer({'max_seq_length': 128, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
)

We can now create some sentences to be turned into sentence embeddings. We'll take some examples from [Marqo's website](https://www.marqo.ai/).

In [3]:
marqo_sentences = [
    "With Marqo you can use your data to increase relevance with embedding search",
    "Join the conversation on our Marqo Community Channel",
    "AI search that understands the way your customers think",
    "How RedBubble increased revenue with Marqo",
    "With Marqo you can use your data to increase downloads with embedding search"
]

embeddings = model.encode(marqo_sentences)

As we’ve seen in the SentenceTransformer object above, we expect these sentence embeddings to be 768-dimensional. Let’s see if they are:

In [4]:
embeddings.shape

(5, 768)

So, yes! We have 5 sentences that are all 768-dimensional. Awesome!

Let’s use cosine similarity to compute how similar each sentence is:

In [5]:
import numpy as np
from sentence_transformers.util import cos_sim

# Assuming `embeddings` is a numpy array or a list of numpy arrays
embeddings = np.array(embeddings)

# Compute the cosine similarity matrix
sim = cos_sim(embeddings, embeddings).numpy()

print(sim)

[[0.9999999  0.6356604  0.66937524 0.70119    0.9407738 ]
 [0.6356604  0.9999999  0.6081568  0.32855833 0.58871996]
 [0.66937524 0.6081568  1.         0.36581916 0.59990084]
 [0.70119    0.32855833 0.36581916 1.         0.74803436]
 [0.9407738  0.58871996 0.59990084 0.74803436 1.0000001 ]]


We can see that sentence 1 and sentence 5 are similar as the (5, 1) entry (bottom-left) is `0.9407738`. Take a glance yourself at the different values and come to a conclusion over which sentences you think are similar.

In this example we used SBERT but there are so many other sentence transformer models available. After the release of SBERT, newer and faster models were released and as a result, significantly outperformed the original SBERT. Interestingly, SBERT is actually no longer listed as an available model on SBERT.net’s models page!

Here are some of the best SBERT models currently available:

- `all-mpnet-base-v2`
- `all-MiniLM-L6-v2`
- `all-distilroberta-v1`
- `paraphrase-mpnet-base-v2`

Feel free to change your code to use these models and see what results you get!

In the next tutorial, we’ll explore some of these newer models in more detail and explain how you can train and fine-tune your own sentence transformers!